lesson20.md

```diff
@@ -0,0 +1,319 @@
+    instructor: true
+};
+
+// turn our object to JSON:
+let json = JSON.stringify(obj);
+
+// turn our JSON back to an object:
+let jsonAsObj = JSON.parse(json);
+```
+
+---
+
+### Quiz
+
+  What does the following output?
+
+```js
+typeof JSON.stringify({ name: "Bob" });
+typeof JSON.parse('{"city":"Berlin"}');
+```
+
+---
+
+### Quiz - answer
+
+```js
+typeof JSON.stringify({ name: "Bob" }); // "string"
+typeof JSON.parse('{"city":"Berlin"}'); // "object"
+```
+
+---
+
+### Web API
+
+* API stands for **A**pplication **P**rogramming **I**nterface
+* A Web API is the *contract* between your application and a server
+
+![API](images/API.png) <!-- .element height="400px" width="400px" -->
+
+---
+
+### REST API
+
+* REST stands for **Re**presentational **S**tate **T**ransfer
+* REST APIs work via http(s), the same protocol we use to connect to any website
+* REST APIs commonly use JSON to exchange data
+* Example for a REST API URL: https://restcountries.eu/rest/v2/lang/de
+* More info: https://developer.mozilla.org/en-US/docs/Glossary/REST
+
+---
+
+### What is missing?
+
+* Now we know the data format to **exchange** data with a server: JSON
+* Now we know the **interface** to a server: REST API
+* But how do we **connect** to the server from within JavaScript?
+
+---
+
+### The dark past: XMLHttpRequest
+
+```js
+let url = 'https://something.com/';
+
+function load(url, callback) {
+ let xhr = new XMLHttpRequest();
+
+ xhr.onreadystatechange = function() {
+   if (xhr.readyState === 4) {
+     callback(xhr.response);
+   }
+ }
+
+ xhr.open('GET', url, true);
+ xhr.send(null);
+}
+```
+
+---
+
+### fetch - Promises
+
+```js
+fetch("http://something.com").then(function(response) { return response.json(); }).then(function(json) { console.log("got json: " + json); });
+```
+
+---
+
+### fetch - async/await
+
+```js
+async function fetchMyApi() {
+  let response = await fetch("http://something.com");
+  let result = await response.json();
+  console.log(result);
+}
+```
+
+---
+
+### async/await
+
+* Fetching data from the internet might take a lot of time
+* We must not block our browser while waiting for the reply!
+* `XMLHttpRequest` was working with callbacks. The API is ugly, error-prone and difficult to use
+* `fetch` is a replacement. It returns a `Promise` object and doesn't require callbacks.
+* `fetch` with `await/async` "hides" the complexity of `Promise`.
+
+---
+
+### async/await
+
+* JavaScript introduced two new keywords: `async` and `await`
+* We can use the `await` keyword only in `async` functions!
+* `await` executes the call, continues executing the code which comes after the `async` function
+* once the result of the `await` is available, the code in the `async` function resumes
+* This allows us to execute long-running functions without blocking the browser
+
+---
+
+### Full fetch example
+
+```js
+async function fetchMyApi() {
+  // connects to something.com
+  let response = await fetch("http://something.com");
```

Comment:
Maybe explain the nature of two step process, by illuminating that fetch does not contain the body of the HTTP response, only when we call the .json() we make the actual call to get the json formatted data and convert it from string to JS Object.